org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/form/WSQLQuery.java

```diff
@@ -0,0 +1,318 @@
+		layout.appendChild(north);
+
+		// create the bottom row of components
+		m_txbResultField.setRows(noResultRows);
+		ZKUpdateUtil.setHflex(m_txbResultField, "1");
+		m_txbResultField.setReadonly(true);
+
+		center.appendChild(listbox);
+		ZKUpdateUtil.setVflex(listbox, "1");
+		ZKUpdateUtil.setHflex(listbox, "1");
+
+		layout.appendChild(center);
+
+		south.appendChild(m_txbResultField);
+		layout.appendChild(south);
+
+		this.appendChild(layout);
+
+		return;
+	}
+
+	/**
+	 *  Create Process Button.
+	 *  @return button
+	 */
+	public static final Button createProcessButton() {
+		Button btnProcess = new Button();
+		if(ThemeManager.isUseFontIconForImage())
+			btnProcess.setIconSclass("z-icon-Process");
+		else
+			btnProcess.setImage(ThemeManager.getThemeResource("images/Process24.png"));
+		btnProcess.setName(Msg.getMsg(Env.getCtx(), "Process"));
+
+		return btnProcess;
+	}   //  createProcessButton
+
+	/**
+	 *  Process SQL Statements.
+	 *
+	 *  @param sqlStatement a single SQL statement
+	 *  @param allowDML     whether to allow DML statements
+	 *  @return a string summarizing the results
+	 */
+	public String processStatement (String sqlStatement) {
+		m_txbResultField.setText(null);
+		listbox.clear();
+
+		if (sqlStatement == null || sqlStatement.length() == 0)
+			return "";
+		StringBuilder sb = new StringBuilder();
+		char[] chars = sqlStatement.toCharArray();
+		for (int i = 0; i < chars.length; i++) {
+			char c = chars[i];
+			if (Character.isWhitespace(c))
+				sb.append(' ');
+			else
+				sb.append(c);
+		}
+		String sql = sb.toString().trim();
+		if (sql.length() == 0)
+			return "";
+		//
+		StringBuilder result = new StringBuilder();
+		String SQL = sql.toUpperCase();
+		String cleanSQL = SQL
+				.replaceAll(REGEX_REMOVE_COMMENTS, "")
+				.replaceAll(REGEX_REMOVE_QUOTED_STRINGS, "")
+				.replaceFirst(REGEX_REMOVE_LEADING_SPACES, "");
+
+		if (cleanSQL.contains(";")) {
+			result.append("ERROR: Multiple Commands Not Allowed");
+			return result.toString();
+		}
+
+		int timeout = MSysConfig.getIntValue(MSysConfig.FORM_SQL_QUERY_TIMEOUT_IN_SECONDS, 120);
+		int maxRecords = MSysConfig.getIntValue(MSysConfig.FORM_SQL_QUERY_MAX_RECORDS, 500);
+		String[] allowedKeywords = MSysConfig.getValue(MSysConfig.FORM_SQL_QUERY_ALLOWED_KEYWORDS, "SELECT ,WITH ,SHOW ").split(",");
+		boolean isError = true;
+		for (int i = 0; i < allowedKeywords.length; i++) {
+			if (cleanSQL.startsWith(allowedKeywords[i])) {
+				isError = false;	
+				break;
+			}
+		}
+		if (isError) {
+			result.append("ERROR: Not Allowed Command");
+			return result.toString();
+		}
+
+		List<String> header = new ArrayList<String>();
+		header.add("#");
+		model = new ListModelTable();
+		Frozen frozen = new Frozen();
+		frozen.setColumns(1);
+		listbox.appendChild(frozen);
+
+		PreparedStatement pstmt = null;
+		ResultSet rs = null;
+		try {
+			long start = System.currentTimeMillis();
+			pstmt = DB.prepareNormalReadReplicaStatement(sql, null);
```

Comment:
This is a potential point of confusion for user, it is not mention anywhere in the form that it is reading from the read replica DB.

Reply:
Hi @hengsin - there is no difference for the user, the reading from replica is guaranteed to be exact, I think there is no source of confusion because of that.  Or, do you mean something different?

Reply:
that's configuration dependent, if it is async replica, you can have delay between master and replica. I believe it is always better to be more transparent here.

Reply:
That's controlled in our code, otherwise we could not guarantee that a report is accurate with latest changes:

See https://wiki.idempiere.org/en/NF6.2_Reporting_from_Read-Only_Replica

Before running the query we verify if the async replica is in sync, if not it waits and iterate, until a max of iterations and as last resort it falls back to master.